### DISCUSSION TITLE
Preforming a redirect from a middleware

### DISCUSSION THREAD

**Discussion by vendion** [goravel/goravel#286] [Q&A]
In building a small test web app that makes use of cookies to store a user's auth token, and an HTTP middleware to check if that cookie exists and validates the value using `facades.Auth().Parse()`. Going down this route I have a start, and when the cookie exists and the token is valid, all is good. The problem is if the cookie doesn't exist/is no longer valid, I want to redirect the user to the login screen, but that I can't seem to get working.

```go
package auth

import (
        "goravel/app/models"

        "github.com/goravel/framework/contracts/http"
        "github.com/goravel/framework/facades"
)

func IsAuthed() http.Middleware {
        return func(ctx http.Context) {
                c, err := ctx.Request().Origin().Cookie("session_token")
                if err != nil || c.String() == "" {
                        facades.Log().Debug("foo")
                        ctx.Response().Redirect(http.StatusUnauthorized, "/")
                        return
                }

                facades.Log().Debugf("cookie: %#v", c.String())

                if err = c.Valid(); err != nil || c.Value == "" {
                        ctx.Response().Redirect(http.StatusUnauthorized, "/")
                }

                _, err = facades.Auth().Parse(ctx, c.Value)
                if err != nil {
                        ctx.Response().Redirect(http.StatusUnauthorized, "/")
                }

                var user models.User
                if err := facades.Auth().User(ctx, &user); err != nil {
                        ctx.Response().Redirect(http.StatusUnauthorized, "/")
                }

                ctx.WithValue("user", user)
                ctx.Request().Next()
        }
}
```

Using the log, I can see that the `if err != nil || c.String() == ""` is executed, but rather than getting redirected once `return` gets called, the request continues on as if `ctx.Request().Next()` got called.

Using `ctx.Request().AbortWithStatus()` does prevent the request from continuing, but given my use case, using Goravel for the backend and frontend, does not give the expected results. Instead, it just returns the HTTP status code and no data for the browser to render, making the site appear broken.

**Comment by hwbrzzl** [comment]
You can try this for now: `err := ctx.Response().Redirect(http.StatusUnauthorized, "/").Render()`

**Reply by vendion** [reply]
The problem with that is `ctx.Response().Redirect()` doesn't allow redirecting with a 401 status code.

**Reply by hwbrzzl** [reply]
Sorry, not get your meaning.

**Reply by vendion** [reply]
Take this code for example:

```go
package middleware

import (
	"github.com/goravel/framework/contracts/http"
	"github.com/goravel/framework/facades"
)

// Auth checks if the request came from a logged in user.
func Auth() http.Middleware {
	return func(ctx http.Context) {
		t := ctx.Request().Cookie("auth_token")

		payload, err := facades.Auth(ctx).Parse(t)
		if err != nil {
			facades.Log().Errorf("failed to parse user auth_token: %v", err)

			ctx.Response().Redirect(http.StatusUnauthorized, "/").Render()
		}

		facades.Log().Infof("user auth_token payload: %#v", payload)

		ctx.Request().Next()
	}
}
```

The redirect will actually panic

![image](https://github.com/user-attachments/assets/608214f3-f764-42c4-9258-c4b279bb5b11)

**Reply by hwbrzzl** [reply]
Based on [the gin limitation](https://github.com/gin-gonic/gin/blob/e46bd521859fdfc83c508f1d42c92cb7f91e9fcb/render/redirect.go#L21), the code should `Code >= http.StatusMultipleChoices && Code > http.StatusPermanentRedirect`, and should add `ctx.Request().AbortWithStatus(http.StatusMovedPermanently)`:

```go
if err != nil {
  facades.Log().Errorf("failed to parse user auth_token: %v", err)
  
  ctx.Response().Redirect(http.StatusMovedPermanently, "/").Render()
  ctx.Request().AbortWithStatus(http.StatusMovedPermanently)
}
```

**Reply by hwbrzzl** [reply]
And we will optimize the code in v1.15:

```go
-- ctx.Response().Redirect(http.StatusMovedPermanently, "/").Render()
-- ctx.Request().AbortWithStatus(http.StatusMovedPermanently)
++ ctx.Response().Redirect(http.StatusMovedPermanently, "/").Abort()
```

**Comment by istiyakamin** [comment]
ctx.Response().Redirect() only sets the response headers; it doesn’t stop the middleware chain.
After the redirect is written Goravel will keep executing the remaining middleware/handler unless you explicitly abort the request.

So, in every early-exit path:

Use a real redirect status (30x) – a browser will not follow a 401.

Call Abort() (or ctx.Request().Abort()) so nothing else runs.
